### ISSUE TITLE
Migrate to TypeScript

### ISSUE THREAD

**Issue by segunadebayo** [chakra-ui/chakra-ui#205]
We're currently migrating the codebase to TypeScript to provide better type safety across all components.

If you have any typescript related concerns, kindly comment below and find a walkaround for now till the TS version is released.

Cheers.

PS: Want to contribute to the migration, check out the `dev-ts` branch.

**Comment by cschroeter** [comment]
The inferface for `<Switch />` requires a children - which is not intended i guess ^^

**Comment by Vynlar** [comment]
I am tinkering with the new `dev-ts` branch and have an observation: In the `CloseButton` component, the `aria-label` prop is optional; however, I believe that it should be required for accessibility reasons.

The larger question is: should we use the power of types to _enforce_ accessibility? This is what the `elm-ui` library does over in the `elm` world to great effect. It's a bit of a philosophical question and I'd like to see some discussion about it.

Great work! 😄

**Comment by segunadebayo** [comment]
@Vynlar. All components will be reviewed and tested multiple times before we release it. Thanks for that observation though. I will correct that.

I agree that we should use the power of types (for TS) and `console.warn` (for JS) to enforce accessibility. If there's any accessibility we can't handle out of the box, then it makes sense to enforce it and put some nice comments/links to help educate the user.

**Comment by alexluong** [comment]
Would like to second #194. Not sure what the work around is at the moment.

**Comment by noahcoomer** [comment]
`AvatarGroup` is missing a type definition for the `max` prop

**Comment by brentguistwite** [comment]
@segunadebayo I would love to help contribute to migrating to TS. It's not immediately clear to me after looking through the dev-ts branch where to get started. Is there a place where I can view what still needs to be done to complete the migration?

**Comment by segunadebayo** [comment]
Hi @brentguistwite. Really happy to know you want to contribute. I'll create a quick guide soon

**Comment by koolamusic** [comment]
@segunadebayo looking forward to that quick guide as well.

**Comment by matepapp** [comment]
I think it might be a good solution to use Github's Project feature, where you can track the task, subtasks separately and anyone who wants to contribute can pick up one.

**Comment by segunadebayo** [comment]
@matepapp, that's a good idea! I'll try to set it up in my free time.

**Comment by jordiup** [comment]
I'm getting the following error when using Avatar 

`TypeError: Cannot read property '12' of undefined`

the code: 

`<Avatar name="Kent C" src="https://bit.ly/kent-c-dodds" />`

**Comment by segunadebayo** [comment]
I've added the `TS_MIGRATION_GUIDE.md` to the `dev-ts` branch. It's not complete yet but will give a clear sense of what we're doing and how to get started.

**Comment by koolamusic** [comment]
Seen the TS Guide, the Section that has Components and Todos, should we expect it to be filled up ?

**Comment by segunadebayo** [comment]
@koolamusic, yes I'll complete that section today.

**Comment by riccoski** [comment]
Nice work all. I'm finding the 'as' prop not working with TS. Specifically the Link component.

**Comment by lesha1201** [comment]
@segunadebayo I'm working now on props for `<Box />`. I want to make this component to accept `SystemProps` and props based on element type from `as` prop. I think I finished types for `<Box />` but I need to update types of all components that use `<Box />`. It should fix https://github.com/chakra-ui/chakra-ui/issues/194. I hope to make PR tomorrow if everything goes well

**Comment by brentguistwite** [comment]
> 
> 
> Nice work all. I'm finding the 'as' prop not working with TS. Specifically the Link component.

We should probably avoid using `as` if possible. You lose some type safety when doing type assertions like that. The solution is usually more concise/correct types or if the type is polymorphic then using a type check like `if ('field' in type)` or using a predicate function like 
```
const isType = (foo): foo is bar => {
  // return !!foo.field
}
```

**Comment by segunadebayo** [comment]
@lesha1201, sounds good. I also have a solution for the `Box` and `as` prop but I'm open to your solution. Can't wait to see it.

**Comment by jmsheff** [comment]
> I think it might be a good solution to use Github's Project feature, where you can track the task, subtasks separately and anyone who wants to contribute can pick up one.

What's the status of this? I'm very interested in contributing but I'm having a hard time figuring out what's going on where. If it's a matter of the tediousness of setting up the project feature to get things organized, I'd be happy to work on that.

**Comment by lesha1201** [comment]
@segunadebayo I won't be able to make PR today because I stuck and not sure if I can solve the problem so if you have solution when please don't wait me

**Comment by segunadebayo** [comment]
I've added the Project following @matepapp's suggestion. Want to contribute? Convert any task to an issue to indicate your interest.

I'll try to keep the board up to date.

**Comment by jaredpalmer** [comment]
+1 for Box as. My suggestion is to `infer` or use a JSX generic

**Comment by riccoski** [comment]
For a temporary workaround for using as with the component's props.
```
<MenuItem as={Link} {...{ to: "/logout" }}>
    Log out
</MenuItem>
```

**Comment by Andreeh** [comment]
Im getting `JSX element type 'SimpleGrid' does not have any construct or call signatures.  TS2604` with the latest and greatest

**Comment by sciolist** [comment]
ListItem is a PseudoBox, but only allows BoxProps in its type definition, so you can't use _hover etc on it with typescript.

Can be worked around with an ambient file containing:
```
import { ListItem } from '@chakra-ui/core/dist/List';
import { PseudoBoxProps } from '@chakra-ui/core/dist/PseudoBox';
declare module '@chakra-ui/core' {
    export const ListItem: React.FC<PseudoBoxProps>;
}
```

**Comment by NathanH81** [comment]
Hello,

Thank you for this awesome library. When creating a custom icon via theme, the name property for the Icon component causes a compile error in TypeScript. Relevant CodeSandbox is located here: 

https://codesandbox.io/s/chakra-ts-starter-wr85o

**Comment by leetmachine** [comment]
upgraded from core 0.4.1 to 0.5.1 and my theme.ts file broke due to what looks like a change in ITheme from StyledSystem.Theme to a chakra's own type.

Current fix:
don't use the ITheme

Usage:
I'd like to be able to extend the theme with my own names for keys. My designer used Figma to create the website design and her naming convention does not (and should not) follow Chakra's spec. By allowing my own keys such as shadows.light , I was able to both keep the theme file aligned to the figma design, as well as use it in my component <Box shadow="light />. It feels really clean and extendable to our team style, really liked that.

example of what worked and then broke:
```
shadows: {
    lighter: '0px 2px 4px rgba(87, 87, 87, 0.08)',
    light: '0px 4px 8px rgba(87, 87, 87, 0.16)',
    medium: '0px 16px 24px rgba(87, 87, 87, 0.2)',
  }
```

Other considerations:
Colors - colors is strict on the ColorHues array. Our design is only using a few shades of blue. TS throws an error that we are missing the values for 300, 500, etc. It's an easy work around to spread the theme.colors.blue, but this makes those colors available in the project and we don't want developers using those. Better to make the hue keys optional. 


Really happy with the library overall, appreciate all your efforts!

**Comment by segunadebayo** [comment]
@leetmachine. I'll release a patch for this shortly.

**Comment by sirluis** [comment]
Hey. Wanna help, just saw the types for the Toast component on the project board on To-Do, but the types are already on master branch, where can I help?

**Comment by segunadebayo** [comment]
Hi @sirluis. Thanks for wanting to help.

Here are some tasks you can help with:
- Write tests for some chakra-ui hooks in the `dev-ts` branch.
- Convert the useToast hook in `master` to TS. Currently, in the master branch, we have the types in `index.d.ts` but now we need to merge both into the same file.
- In the `useSlider` hook in `dev-ts`, add support for `multi-thumb` sliders.

Let me know which of these you're looking to solve so we don't duplicate efforts.

**Comment by sirluis** [comment]
Hi @segunadebayo .
Better I can do ATM is write tests. Are you guys testing with jest?

**Comment by segunadebayo** [comment]
Hi @sirluis, yes we have jest and `testing-library` in the repo. feel free to check the `dev-ts` branch to start some work.

There's already a PR for the test here: #282. You can check that as well.

Thanks for choosing to collaborate on this

**Comment by knguyen0125** [comment]
PseudoBox's `_after` and `_before` doesn't have definitions for CSS `content` property, which would require users to annotate `// @ts-ignore` to be able to keep Typescript from complaining.

It can be fixed by changing the definition from

```ts
_after?: BoxProps;

_before?: BoxProps;
```

to 

```ts
_after?: BoxProps & {
    content?: String
}

_before?: BoxProps & {
    content?: String
}
```

**Comment by lorenzorapetti** [comment]
Hi! I was trying to build the library with `yarn build` but it gives me this error:

```
lerna ERR! yarn build stderr:
(typescript) Error: /Users/lorenzo/code/forks/chakra-ui/packages/Button/src/ButtonGroup.tsx(1,31): semantic error TS7016: Could not find a declaration file for module '@chakra-ui/layout'. '/Users/lorenzo/code/forks/chakra-ui/node_modules/@chakra-ui/layout/dist/index.js' implicitly has an 'any' type.
  Try `npm install @types/chakra-ui__layout` if it exists or add a new declaration (.d.ts) file containing `declare module '@chakra-ui/layout';`
Error: /Users/lorenzo/code/forks/chakra-ui/packages/Button/src/ButtonGroup.tsx(1,31): semantic error TS7016: Could not find a declaration file for module '@chakra-ui/layout'. '/Users/lorenzo/code/forks/chakra-ui/node_modules/@chakra-ui/layout/dist/index.js' implicitly has an 'any' type.
  Try `npm install @types/chakra-ui__layout` if it exists or add a new declaration (.d.ts) file containing `declare module '@chakra-ui/layout';`
    at error (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup/dist/rollup.js:5365:30)
    at throwPluginError (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup/dist/rollup.js:11725:12)
    at Object.error (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup/dist/rollup.js:12694:24)
    at Object.error (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup/dist/rollup.js:11924:38)
    at RollupContext.error (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:17194:30)
    at /Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:24958:23
    at arrayEach (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:534:11)
    at forEach (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:9360:14)
    at printDiagnostics (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:24931:5)
    at Object.transform (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:27374:17)
```

The package from which the error is thrown changes every time, I'm guessing it's because lerna build the packages in parallel and some package depends on others to be built first?

Moreover, I need to add the corresponding `export * from` to `Popover.tsx`, `Select.tsx`, `Menu.tsx` and `Tabs.tsx`, otherwise it gives me this error:

```
@chakra-ui/select: rpt2: options error TS6053: File '/Users/lorenzo/code/forks/chakra-ui/packages/Select/src/Select.tsx' not found.
@chakra-ui/popover: rpt2: options error TS6053: File '/Users/lorenzo/code/forks/chakra-ui/packages/Popover/src/Popover.tsx' not found.
@chakra-ui/menu: rpt2: options error TS6053: File '/Users/lorenzo/code/forks/chakra-ui/packages/Menu/src/Menu.tsx' not found.
@chakra-ui/tabs: rpt2: options error TS6053: File '/Users/lorenzo/code/forks/chakra-ui/packages/Tabs/src/Tabs.tsx' not found.
```

Does any of you get the same errors?

**Comment by felipesousa** [comment]
@segunadebayo I think the better solution to get more feedbacks about some issue or something like that, can be created a document or some other thing to track better!

**Comment by BadMask121** [comment]
I noticed that all the components have no written test, i started investigating on this following the following Reddit post, 

https://www.reddit.com/r/reactjs/comments/d013i8/chakra_ui_40_composable_and_accessible_components/fdrsjwk/


I feel it's important for a huge project like chakra tests should be written for any new features and already existing features.

**Comment by hawk-ee** [comment]
> I noticed that all the components have no written test, i started investigating on this following the following Reddit post,
> 
> https://www.reddit.com/r/reactjs/comments/d013i8/chakra_ui_40_composable_and_accessible_components/fdrsjwk/
> 
> I feel it's important for a huge project like chakra tests should be written for any new features and already existing features.

Well, it's open source, feel free to pick a component and open a PR with tests written. No one is obligated to use it ; )

**Comment by jaredpalmer** [comment]
@segunadebayo just discovered workaround for current `as` prop TS annoyance in reach-ui source. This gem adds full typesafety to stuff like `<Button as={Link} isExternal />`. 

https://gist.github.com/jaredpalmer/80982b3d787359762506fce578108358

In another PR, someone just needs to go through all `forwardRefs()` and replace with `forwardRefWithAs()` in .d.ts files

**Comment by kripod** [comment]
@jaredpalmer Recently, I published a package which serves the exact same purpose, but without the overhead of the `forwardRefWithAs()` wrapper function. It’s called [react-polymorphic-box](https://github.com/kripod/react-polymorphic-box) and offers a set of performant static types along with a guide.

**Comment by ivan-dalmet** [comment]
> Hi! I was trying to build the library with `yarn build` but it gives me this error:
> 
> ```
> lerna ERR! yarn build stderr:
> (typescript) Error: /Users/lorenzo/code/forks/chakra-ui/packages/Button/src/ButtonGroup.tsx(1,31): semantic error TS7016: Could not find a declaration file for module '@chakra-ui/layout'. '/Users/lorenzo/code/forks/chakra-ui/node_modules/@chakra-ui/layout/dist/index.js' implicitly has an 'any' type.
>   Try `npm install @types/chakra-ui__layout` if it exists or add a new declaration (.d.ts) file containing `declare module '@chakra-ui/layout';`
> Error: /Users/lorenzo/code/forks/chakra-ui/packages/Button/src/ButtonGroup.tsx(1,31): semantic error TS7016: Could not find a declaration file for module '@chakra-ui/layout'. '/Users/lorenzo/code/forks/chakra-ui/node_modules/@chakra-ui/layout/dist/index.js' implicitly has an 'any' type.
>   Try `npm install @types/chakra-ui__layout` if it exists or add a new declaration (.d.ts) file containing `declare module '@chakra-ui/layout';`
>     at error (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup/dist/rollup.js:5365:30)
>     at throwPluginError (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup/dist/rollup.js:11725:12)
>     at Object.error (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup/dist/rollup.js:12694:24)
>     at Object.error (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup/dist/rollup.js:11924:38)
>     at RollupContext.error (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:17194:30)
>     at /Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:24958:23
>     at arrayEach (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:534:11)
>     at forEach (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:9360:14)
>     at printDiagnostics (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:24931:5)
>     at Object.transform (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:27374:17)
> ```
> 
> The package from which the error is thrown changes every time, I'm guessing it's because lerna build the packages in parallel and some package depends on others to be built first?
> 
> Moreover, I need to add the corresponding `export * from` to `Popover.tsx`, `Select.tsx`, `Menu.tsx` and `Tabs.tsx`, otherwise it gives me this error:
> 
> ```
> @chakra-ui/select: rpt2: options error TS6053: File '/Users/lorenzo/code/forks/chakra-ui/packages/Select/src/Select.tsx' not found.
> @chakra-ui/popover: rpt2: options error TS6053: File '/Users/lorenzo/code/forks/chakra-ui/packages/Popover/src/Popover.tsx' not found.
> @chakra-ui/menu: rpt2: options error TS6053: File '/Users/lorenzo/code/forks/chakra-ui/packages/Menu/src/Menu.tsx' not found.
> @chakra-ui/tabs: rpt2: options error TS6053: File '/Users/lorenzo/code/forks/chakra-ui/packages/Tabs/src/Tabs.tsx' not found.
> ```
> 
> Does any of you get the same errors?

Same error for me :(

**Comment by ivan-dalmet** [comment]
> Hi! I was trying to build the library with `yarn build` but it gives me this error:
> 
> ```
> lerna ERR! yarn build stderr:
> (typescript) Error: /Users/lorenzo/code/forks/chakra-ui/packages/Button/src/ButtonGroup.tsx(1,31): semantic error TS7016: Could not find a declaration file for module '@chakra-ui/layout'. '/Users/lorenzo/code/forks/chakra-ui/node_modules/@chakra-ui/layout/dist/index.js' implicitly has an 'any' type.
>   Try `npm install @types/chakra-ui__layout` if it exists or add a new declaration (.d.ts) file containing `declare module '@chakra-ui/layout';`
> Error: /Users/lorenzo/code/forks/chakra-ui/packages/Button/src/ButtonGroup.tsx(1,31): semantic error TS7016: Could not find a declaration file for module '@chakra-ui/layout'. '/Users/lorenzo/code/forks/chakra-ui/node_modules/@chakra-ui/layout/dist/index.js' implicitly has an 'any' type.
>   Try `npm install @types/chakra-ui__layout` if it exists or add a new declaration (.d.ts) file containing `declare module '@chakra-ui/layout';`
>     at error (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup/dist/rollup.js:5365:30)
>     at throwPluginError (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup/dist/rollup.js:11725:12)
>     at Object.error (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup/dist/rollup.js:12694:24)
>     at Object.error (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup/dist/rollup.js:11924:38)
>     at RollupContext.error (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:17194:30)
>     at /Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:24958:23
>     at arrayEach (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:534:11)
>     at forEach (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:9360:14)
>     at printDiagnostics (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:24931:5)
>     at Object.transform (/Users/lorenzo/code/forks/chakra-ui/node_modules/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js:27374:17)
> ```
> 
> The package from which the error is thrown changes every time, I'm guessing it's because lerna build the packages in parallel and some package depends on others to be built first?
> 
> Moreover, I need to add the corresponding `export * from` to `Popover.tsx`, `Select.tsx`, `Menu.tsx` and `Tabs.tsx`, otherwise it gives me this error:
> 
> ```
> @chakra-ui/select: rpt2: options error TS6053: File '/Users/lorenzo/code/forks/chakra-ui/packages/Select/src/Select.tsx' not found.
> @chakra-ui/popover: rpt2: options error TS6053: File '/Users/lorenzo/code/forks/chakra-ui/packages/Popover/src/Popover.tsx' not found.
> @chakra-ui/menu: rpt2: options error TS6053: File '/Users/lorenzo/code/forks/chakra-ui/packages/Menu/src/Menu.tsx' not found.
> @chakra-ui/tabs: rpt2: options error TS6053: File '/Users/lorenzo/code/forks/chakra-ui/packages/Tabs/src/Tabs.tsx' not found.
> ```
> 
> Does any of you get the same errors?

Same problem here :(

When I trying to remove the `--parallel` in the `build` script I'm getting other errors because `@chakra-ui/system` or `@chakra-ui/utils` are not in packages dependencies of the components. So Lerna can't know how to build in the right sequence.

@segunadebayo any reason to not declare the local dependencies in the package.json of each component?

**Comment by with-heart** [comment]
Hi @segunadebayo, I'd like to jump in and help with this. What can I work on to be most helpful?

**Comment by segunadebayo** [comment]
@jaredpalmer, thank you for that gist. It's really helpful.

I'll use it in the upcoming TS version. I'm happy to merge any PR that implements the `forwardRefWithAs` into the current master.

**Comment by segunadebayo** [comment]
@ivan-dalmet @lorenzorapetti , I've added the local dependencies for most components, I'll do a double check again to be sure all packages have the correct dependencies. I understand this is critical for lerna to work correctly.

**Comment by segunadebayo** [comment]
@with-heart, I need help with the ReadMe and Testing for all components. Kindly take a look and let me know which you can contribute to. I appreciate this.

Feel free to send me a DM on Twitter as well, if you have any questions.

**Comment by aaronmcadam** [comment]
@segunadebayo What do you think of @kripod's solution? It's quite elegant I think.

**Comment by segunadebayo** [comment]
@aaronmcadam, I think @kripod's and @jaredpalmer's solution are both great! I'm currently looking to sync both solutions and add it to the core of Chakra UI.

**Comment by santialbo** [comment]
I'm following this thread with a lot of attention. One question that I have is what kind of breaking changes should we expect when this is out? Are any components/helpers being deprecated?

**Comment by aaronmcadam** [comment]
@santialbo I too am hoping there will be some user guides on the new TypeScript support and any other migration steps we'll need to follow.

**Comment by segunadebayo** [comment]
@santialbo and @aaronmcadam. There'll definitely be a detailed migration guide for the new TypeScript support. No worries 😄

**Comment by shubham-paliwal-fareye** [comment]
I wanted to work in this repo.
Not able to configure dev-ts branch locally.

`(typescript) Error: /chakra-ui/packages/icons/src/index.tsx(24,15): semantic error TS2307: Cannot find module './ArrowForwardIcon'.`

Resolve this issue by changing the file name from **ArrowForwardICon to ArrowForwardIcon.**
How to run a storybook or doc so that I can do the implementation locally.
Please provide the setup guidelines so that I can help.

**Comment by aaronmcadam** [comment]
@shubham-paliwal-fareye I ran into issues running the branch too, which were resolved here: #473

**Comment by shubham-paliwal-fareye** [comment]
> @shubham-paliwal-fareye I ran into issues running the branch too, which were resolved here: #473

Thanks it works.

**Comment by tsaiDavid** [comment]
@segunadebayo - I would love to help out here, anything specific I might be able to help with?

**Comment by segunadebayo** [comment]
@tsaiDavid Amazing! Your help is welcome, I wrote down some tasks I need to finish up for this package but didn't get the time to complete it. Here's it:

https://github.com/chakra-ui/chakra-ui/blob/dev-ts/packages/media-query/todo.md

If you have any issues or questions, you're welcome to chat with me here: https://discord.gg/xKuJzF

Thank you.

**Comment by kripod** [comment]
@segunadebayo With regards to typed media query hooks, I have previously made minimalistic, SSR-compatible implementations for them:

- [`useAnimationPreference`](https://github.com/kripod/react-hooks/blob/master/packages/web-api-hooks/src/usePreferredMotionIntensity.ts)
- [`useColorModePreference`](https://github.com/kripod/react-hooks/blob/master/packages/web-api-hooks/src/usePreferredColorScheme.ts)
- [`useViewportSize`](https://github.com/kripod/react-hooks/blob/master/packages/web-api-hooks/src/useSize.ts)

**Comment by segunadebayo** [comment]
Amazing! @kripod. Thanks, buddy

**Comment by remorses** [comment]
Hey @segunadebayo, the `ts-dev` will be impossible to easily merge into master if we don't start gradually merging it into master asap.

We could already merge the progress made in `ts-dev` into master and add the `includeJavascript` field in tsconfig to still keep the components written in jsx.

What do you think?

**Comment by segunadebayo** [comment]
@remorses I had this same thought as well. Thanks for bring it up.

We'll think of a way to start merging or rebasing it with `master`

**Comment by maddnespj** [comment]
Hi, i really like chakra ui, and it's great that you are migrating to typescript. But when do you think there is a stable version released of it? (dev-ts as gone into master and been released to npm)

**Comment by potty** [comment]
Good TS support is crucial for our company. So now I'm waiting for a stable Chakra release, so I can finally migrate from Ant Design. Anyway, great work guys, hope this won't take too long, I'm really excited.

**Comment by yusufali19** [comment]
While using Checkbox component in my project I'm getting this error 
```
TypeError: right-hand side of 'in' should be an object, got undefined
useVariantColorWarning
node_modules/@chakra-ui/core/dist/es/utils/index.js:113

  110 | var variantColorIsDefined = variantColor != null;
  111 | 
  112 | if (variantColorIsDefined) {
> 113 |   var variantColorExists = variantColor in theme.colors; // If variant color exists in theme object
      | ^  114 | 
  115 |   if (!variantColorExists) {
  116 |     console.warn("You passed an invalid variantColor to the " + label + " Component. Variant color values must be a color key in the theme object that has '100' - '900' color values. Check http://chakra-ui.com/theme#colors to see possible values");
```

**Comment by with-heart** [comment]
Hi @yusufali19, could you please open a separate issue for your problem?

**Comment by with-heart** [comment]
For anyone following this topic, @segunadebayo has posted an update on where we are with this migration and other features in #696. We're shooting to release 1.0 by the end of May, or the end of June at the latest.

**Comment by gabriel-fortin** [comment]
A very similar one was already mentioned:
```
<MenuButton as={Button} rightIcon="chevron-down" >
    Action
</MenuButton>
```
The `rightIcon` prop is not recognised.
Using `{...{rightIcon: "chevron-down"}}` as a workaround works well.

I'm using `"@chakra-ui/core": "^0.7.0",`

**Comment by mrmckeb** [comment]
@gabriel-fortin you can also do this for some safety:
```tsx
<MenuButton
  as={Button}
  aria-label='Settings'
  {...{
    rightIcon: 'chevron-down',
    type: 'button',
    variant: 'ghost',
  } as React.ComponentProps<typeof Button>}
>
  Example
</MenuButton>
```

**Comment by with-heart** [comment]
We've had a ton of updates since this post was last updated officially. Please see #906 for the most up-to-date thread on the 1.0 journey!
